[all]: Added AArch64 MVN instruction

lukeg101 opened the pull request.

maranget commented:
`MVN` is a shorthand for the `ORN` instructions with `WZR/XZR` as second argumeng. I know that we have implemented such shortcuts directly in the past, but would'nt it be a goog idea to implement `ORN` with `MVN` as a special case? For an example, the `CMP` instruction is implemented as a special case of the `SUBS` instruction.

lukeg101 commented:
Hi Luc, Thanks again for the review. I have implemented `MVN` as `ORN`

maranget commented:
Apart from the conflict (A231.litmus can be renamed into A236.litmus, also changing occurrences of names A231 into A236 in files A236.litmus and A236.litmus.expected), there are still a few minor problem.
1. It is a pity not to parse the `ORN` instruction as everything is in place to do so
2. ORN with XZR/WZR second argument would better be printed as `MVN` both in pretty-printed output (_i.e._ in `lib/AArch64Base.ml`) and in assembler output (_i.e._ in `litmus/AArch64Compile`).

lukeg101 commented:
Hi Luc, thanks for your help again, I have done the following:
1. renamed test `A231` to `A236`
2. Added test for `ORN`, namely `A237`
3. updated `lib` printing to print `MZN` when `ORN` has the zero reg in the second argument
4. Repeated the same for litmus

all herd tests pass, however litmus does not give the same result (it returns `0:X0 = 0 /\ 0:X3 = 0`) when running the above tests on an Apple M1 machine. Have I missed something in the test?

maranget commented on litmus/AArch64Compile_litmus.ml:
```diff
@@ -1131,6 +1131,24 @@ module Make(V:Constant.S)(C:Config) =
+          and r2,f2 = arg1 "xzr" (fun s -> "^i"^s) r2 in
+          { empty_ins with
+            memo=sprintf "%s %s,%s" memo f1 f2;
+            inputs=r1;
+            outputs=r2; reg_env = add_q (r1@r2);}
```
Here, inputs and outputs have to be swapped:
```suggestion
            inputs=r2;
            outputs=r1; reg_env = add_q (r1@r2);}
```

maranget commented on herd/AArch64Sem.ml:
```diff
@@ -2350,6 +2349,8 @@ module Make
                     | ADD | ADDS -> fun (v1, v2) -> M.add v1 v2
                     | EOR -> fun (v1, v2) -> M.op Op.Xor v1 v2
                     | ORR -> fun (v1, v2) -> M.op Op.Or v1 v2
+                    | ORN -> fun (v1, v2) -> M.op1 Op.Not v2
```
The operation performed by ORN is bitwise NOT, which herd implements as `Op.Inv`.
```suggestion
                    | ORN -> fun (v1, v2) -> M.op1 Op.Inv v2
```

maranget commented on herd/tests/instructions/AArch64/A236.litmus:
```diff
@@ -0,0 +1,9 @@
+P0;
+  MVN X0, X1;
+  MVN X3, X2;
+forall (0:X0 = 1 /\ 0:X3 = 0)
```
Notice output will change to
```suggestion
forall (0:X0 = -1 /\ 0:X3 = -2)
```

maranget commented on herd/tests/instructions/AArch64/A237.litmus:
```diff
@@ -0,0 +1,10 @@
+AArch64 237
+
+(* Tests ORN Register instruction *)
+{ int64_t 0:X1 = 0; int64_t 0:X2 = 1}
```
Same remark as for A236:
```suggestion
{ int64_t 0:X1 = 0; int64_t 0:X2 = 1;
  int64_t 0:X0; int64_t 0:X3; }
```

maranget commented on herd/tests/instructions/AArch64/A237.litmus:
```diff
@@ -0,0 +1,10 @@
+{ int64_t 0:X1 = 0; int64_t 0:X2 = 1}
+
+P0;
+  ORN X0, XZR, X1;
+  ORN X3, XZR, X2;
```
For little effort we can have `orn` in emitted  assembler:
```suggestion
  ORN X0, X1, X1;
  ORN X3, X1, X2;
```

lukeg101 commented:
Negative value fix in https://github.com/herd/herdtools7/pull/510

maranget commented:
I have squashed and merged this PR by hand, also adding some `int64_t` type declarations in tests A236 and A237. For some reason this approach has created a conflict with this branch. However, the PR has been merged,. So I close it.

Thanks @lukeg101 for the implementation of two additional  instructions.
